src/test/test_marshal.cpp

```diff
@@ -316,96 +516,147 @@ TEST_CASE("unmarshal extension opcodes", "[disasm][marshal]") {
         ebpf_inst{.opcode = INST_ALU_OP_MOV | INST_SRC_REG | INST_CLS_ALU64, .dst = 1, .src = 1, .offset = 32});
 }
 
-TEST_CASE("fail unmarshal invalid opcodes", "[disasm][marshal]") {
-    // The following opcodes are undefined and should generate bad instruction errors.
-    uint8_t bad_opcodes[] = {
-        0x00, 0x01, 0x02, 0x03, 0x08, 0x09, 0x0a, 0x0b, 0x0d, 0x0e, 0x10, 0x11, 0x12, 0x13, 0x19, 0x1a, 0x1b, 0x60,
-        0x68, 0x70, 0x78, 0x80, 0x81, 0x82, 0x83, 0x86, 0x88, 0x89, 0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f, 0x90, 0x91,
-        0x92, 0x93, 0x96, 0x98, 0x99, 0x9a, 0x9b, 0x9d, 0x9e, 0xa0, 0xa1, 0xa2, 0xa3, 0xa8, 0xa9, 0xaa, 0xab, 0xb0,
-        0xb1, 0xb2, 0xb3, 0xb8, 0xb9, 0xba, 0xbb, 0xc0, 0xc1, 0xc2, 0xc8, 0xc9, 0xca, 0xcb, 0xd0, 0xd1, 0xd2, 0xd3,
-        0xd8, 0xd9, 0xda, 0xdf, 0xe0, 0xe1, 0xe2, 0xe3, 0xe4, 0xe5, 0xe6, 0xe7, 0xe8, 0xe9, 0xea, 0xeb, 0xec, 0xed,
-        0xee, 0xef, 0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9, 0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff};
-    for (int i = 0; i < sizeof(bad_opcodes); i++) {
-        std::ostringstream oss;
-        oss << "0: Bad instruction op 0x" << std::hex << (int)bad_opcodes[i] << std::endl;
-        check_unmarshal_fail(ebpf_inst{.opcode = bad_opcodes[i]}, oss.str().c_str());
+// Check that unmarshaling an invalid instruction fails with a given message.
+static void check_unmarshal_instruction_fail(ebpf_inst& inst, const char* message) {
+    if (inst.offset == JMP_OFFSET) {
+        inst.offset = 1;
+        check_unmarshal_fail_goto(inst, message);
+    } else if (inst.opcode == INST_OP_LDDW_IMM)
+        check_unmarshal_fail(inst, ebpf_inst{}, message);
+    else
+        check_unmarshal_fail(inst, message);
+}
+
+// Check that various 'dst' variations between two valid instruction templates fail.
+static void check_instruction_dst_variations(const ebpf_inst& previous_template, const ebpf_inst* next_template) {
+    ebpf_inst inst = previous_template;
+    if (inst.dst == DST) {
+        inst.dst = R10_STACK_POINTER + 1; // Not a valid register.
+        check_unmarshal_instruction_fail(inst, "0: Bad register\n");
+    } else {
+        // This instruction doesn't put a register number in the 'dst' field.
+        // Just try the next value unless that's what the next template has.
+        inst.dst++;
+        if (!next_template || memcmp(&inst, next_template, sizeof(inst)) != 0) {
```

Comment:
I'm not sure why memcmp and not a simple comparison. I think it can be easily defined (as a non-member operator), and even defined to use memcmp if needed, though I don't see why.

Additionally, I think that using std::optional instead of a pointer would allow the entire test to be a simple comparison `inst != next_template`.

Reply:
Done